### PR TITLE
make iotools package

### LINKED ISSUE #29: add a io module
From:
https://github.com/pvlib/pvlib-python/pull/7#issuecomment-78441084

Shall we add standard readers for typical data providers such as:

Input:
-  Meteonorm
-  SolarGIS, e.g. http://solargis.info/doc/_files/sample_data/SolarGIS-climData/SolarGIS-TMY-hourly-P50-PSA-ES.zip
- BSRN
- SoDa-IS

Output:
- epw: http://apps1.eere.energy.gov/buildings/energyplus/cfm/weather_data3.cfm
- SAM

Hints:
https://github.com/pydata/pandas/blob/master/pandas/io/date_converters.py

There was a similar effort at pandas some time ago which stalled:
https://github.com/pydata/pandas/issues/1180

This could be a meta issue for all input and output format existing in the (solar) energy meteo domain.

### LINKED ISSUE #261: Design pvlib.io
I'm assembling code to create pvlib.io. 

What functionality is desired?  Specifically, from which formats do we want to read, and to which formats do we want to write?

I'm going to focus first on reading a few data formats. I plan to incorporate the current pvlib.tmy which reads tmy2 and tmy3 files, returning a tuple (data, metadata) where data is a pandas DataFrame and metadata is a dict.  Any comments on this design?  Personally I like it.

I have code to read surfrad files so I'll include that.

### PR DESCRIPTION
**Pull request by wholmgren**
I made a new ``io`` branch to try to streamline the process of multiple users contributing their new io tools and to make it easier for all of us to test it them without hurting the master branch. I also went ahead and moved the tmy readers into the package and updated the tests accordingly. I named the package ``iotools`` since Cliff expressed a preference for that name, but it's not yet set in stone. I think this is mostly consistent with what we talked about in #261 and #29. Let me know if I missed something concerning the structure.

Please make your new IO tools pull requests against this branch rather than master.

Finally, I'm not going to merge into master until we're ready to make a series of changes to the API. I want to do at least 0.4 release with non-API breaking changes before then.

Closes #29. Closes #261.

### CODE REVIEW

**Comment by dacoex** [comment]
I added a reader for PVSyst files:
https://github.com/dacoex/pvlib-python/commits/io
--> [added pvsyst reader](https://github.com/dacoex/pvlib-python/commit/8e29054dc723be7a745ec4b709fbcea32696cb38)

This also shows the challenge in getting metadata from CSV files of different sources.
It also shows that pvlib variable list is not complete when compared to PVSyst output, for instance.

I hope this helps to define
- minimum specifications for the io readers
- optional components of the readers

I think we shall agree on the general structure before improving each reader (documentation / testing)

**Comment by wholmgren** [comment]
> I think we shall agree on the general structure before improving each reader (documentation / testing)

I think we essentially agreed on the general structure in #261. Is there some part of the structure that you disagree with? At this point, I think our time is better spent focusing on implementing new readers.

**Comment by cwhanse** [comment]
@dacoex thanks so much for getting this off the ground. And to @wholmgren for setting up the branch for development.  I'll get the surfrad reader uploaded, and copy the tmy readers over.  Great to see the progress!

**Comment by wholmgren** [comment]
Great! Be sure to fetch and checkout the new io branch from the main github organization before adding your surfrad reader. I already moved the Tmy readers so it shouldn't be necessary to touch them. Sorry Cliff and @dacoex for the git confusion and my poor explanations.

**Comment by wholmgren** [comment]
I think this is ready to be reconsidered after almost 2 years. In short:

* `tmy.readtmy2` becomes `iotools.read_tmy2`
* `tmy.readtmy3` becomes `iotools.read_tmy3`
* new IO related tools will go in new modules under `iotools`, with their top-level functions exposed through `iotools.api`

The old functions remain but emit deprecation warnings. I marked them for removal in 0.7 for now.

Here's a link to the rendered API documentation: https://pvlib-python.readthedocs.io/en/io/api.html#io-tools

We'll make a few PRs for adding U Oregon SRML and Surfrad in the next couple of weeks. We'll make the PRs to the `io` branch if it's not yet merged into master at that point.

I'm ignoring the majority of the Stickler errors since they're not really new and I don't care to take the time to fix them. Unfortunately I don't know how to turn them off for this PR. At least it didn't inline the comments!

**Comment by wholmgren** [comment]
Closes #29. Closes #261.

@mikofski any new thoughts/objections on package structure 2 years later?

**Comment by mikofski** [comment]
TL;DR: I like the new structure. `iotools.read_tmy3` is a lot more elegant than `tmy.read_tmy3` and I like that many different data sources and formats can be grouped together, in addition to TMY. It will be interesting to see what output formats there are in `iotools`, since currently there are only feature requests for inputs.

Although I think `api` is fine, it seems redundant, since IMO that's what `__init__` is for, but I'm not opposed to `api`, I just think it's unnecessary. So if you think it's more explicit or that `__init__` is less readable than that seems fine to me

**Comment by mikofski** [comment]
>[I like the `api.py` and empty `__init__.py` idea](https://github.com/pvlib/pvlib-python/issues/235#issuecomment-260789076)

Sorry, that statement from 2 years ago contradicts my review here. I guess I'm ambivalent. But I respect your judgement

**Comment by wholmgren** [comment]
Hearing no objections, I will go ahead and merge this. We can easily change (or even revert) through the next release.
